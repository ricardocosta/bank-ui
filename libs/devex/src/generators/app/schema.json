{
  "$schema": "http://json-schema.org/schema",
  "cli": "nx",
  "$id": "App",
  "title": "App generator",
  "type": "object",
  "properties": {
    "name": {
      "type": "string",
      "description": "The name of the app",
      "$default": {
        "$source": "argv",
        "index": 0
      },
      "x-prompt": "What name would you like to use?"
    },
    "tags": {
      "type": "string",
      "description": "Add tags to the project (used for linting)"
    },
    "directory": {
      "type": "string",
      "description": "A directory where the project is placed"
    },
    "e2e": {
      "type": "boolean",
      "description": "Whether or not to create an e2e test app",
      "default": true,
      "x-prompt": "Create E2E test app?"
    },
    "restServer": {
      "type": "boolean",
      "description": "Whether or not to bootstrap a REST server using json-server",
      "default": false,
      "x-prompt": "Add development REST server?"
    }
  },
  "required": ["name", "e2e", "restServer"]
}
