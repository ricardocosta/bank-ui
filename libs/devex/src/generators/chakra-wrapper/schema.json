{
  "$schema": "http://json-schema.org/schema",
  "cli": "nx",
  "$id": "ChakraWrapper",
  "title": "Chakra UI Wrapper generator",
  "type": "object",
  "properties": {
    "name": {
      "type": "string",
      "description": "The name of the wrapper (ui-<name>)",
      "$default": {
        "$source": "argv",
        "index": 0
      },
      "x-prompt": "What name would you like to use?"
    },
    "tags": {
      "type": "string",
      "description": "Add tags to the project (used for linting)"
    },
    "chakraPackage": {
      "type": "string",
      "description": "Whether or not to create an e2e test app",
      "x-prompt": "Which ChakraUI package do you want to wrap?"
    }
  },
  "required": ["name", "chakraPackage"]
}
