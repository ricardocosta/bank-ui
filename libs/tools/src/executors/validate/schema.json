{
  "$schema": "http://json-schema.org/schema",
  "cli": "nx",
  "title": "Validate",
  "description": "Validate the project by running ESLint, Prettier and TSC",
  "type": "object",
  "properties": {},
  "required": []
}
