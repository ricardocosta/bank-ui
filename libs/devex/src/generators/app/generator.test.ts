import { readJson, readProjectConfiguration } from "@nrwl/devkit";
import { createTreeWithEmptyWorkspace } from "@nrwl/devkit/testing";
import { describe } from "vitest";

import appGenerator from "./generator";
import { AppGeneratorSchema } from "./schema";

const DEMO_APP_ROOT = "apps/demo";
const DEMO_APP_E2E_ROOT = "apps/demo-e2e";

describe("Generators / App", () => {
  const options: AppGeneratorSchema = { name: "demo", e2e: false, restServer: false };

  it("generates the project configuration", async () => {
    const appTree = createTreeWithEmptyWorkspace({ layout: "apps-libs" });
    await appGenerator(appTree, options);

    const config = readProjectConfiguration(appTree, "demo");

    expect(config).toHaveProperty("$schema", "../../node_modules/nx/schemas/project-schema.json");
    expect(config).toHaveProperty("name", "demo");
    expect(config).toHaveProperty("projectType", "application");
    expect(config).toHaveProperty("sourceRoot", "apps/demo/src");
    expect(config).toHaveProperty("root", "apps/demo");
    expect(config).toHaveProperty("tags", []);
  });

  it("adds targets to the project configuration", async () => {
    const appTree = createTreeWithEmptyWorkspace({ layout: "apps-libs" });
    await appGenerator(appTree, options);

    const config = readProjectConfiguration(appTree, "demo");

    config.targets && expect(Object.keys(config.targets)).toHaveLength(10);
    expect(config.targets).toHaveProperty("build");
    expect(config.targets).toHaveProperty("dev");
    expect(config.targets).toHaveProperty("test");
    expect(config.targets).toHaveProperty("testCoverage");
    expect(config.targets).toHaveProperty("testUi");
    expect(config.targets).toHaveProperty("testWatch");
    expect(config.targets).toHaveProperty("lint");
    expect(config.targets).toHaveProperty("format");
    expect(config.targets).toHaveProperty("formatDry");
    expect(config.targets).toHaveProperty("ts");
  });

  it("generates the application files", async () => {
    const appTree = createTreeWithEmptyWorkspace({ layout: "apps-libs" });
    await appGenerator(appTree, options);

    expect(appTree.exists(`${DEMO_APP_ROOT}/index.html`)).toBeTruthy();
    expect(appTree.exists(`${DEMO_APP_ROOT}/setupTests.ts`)).toBeTruthy();
    expect(appTree.exists(`${DEMO_APP_ROOT}/tsconfig.json`)).toBeTruthy();
    expect(appTree.exists(`${DEMO_APP_ROOT}/tsconfig.node.json`)).toBeTruthy();

    expect(appTree.exists(`${DEMO_APP_ROOT}/src/App.css`)).toBeTruthy();
    expect(appTree.exists(`${DEMO_APP_ROOT}/src/App.test.tsx`)).toBeTruthy();
    expect(appTree.exists(`${DEMO_APP_ROOT}/src/App.tsx`)).toBeTruthy();
    expect(appTree.exists(`${DEMO_APP_ROOT}/src/favicon.svg`)).toBeTruthy();
    expect(appTree.exists(`${DEMO_APP_ROOT}/src/index.css`)).toBeTruthy();
    expect(appTree.exists(`${DEMO_APP_ROOT}/src/logo.svg`)).toBeTruthy();
    expect(appTree.exists(`${DEMO_APP_ROOT}/src/main.tsx`)).toBeTruthy();
    expect(appTree.exists(`${DEMO_APP_ROOT}/src/vite-env.d.ts`)).toBeTruthy();
  });

  it("adds ESLint configuration", async () => {
    const appTree = createTreeWithEmptyWorkspace({ layout: "apps-libs" });
    await appGenerator(appTree, options);

    expect(appTree.exists(`${DEMO_APP_ROOT}/.eslintrc.js`)).toBeTruthy();
  });

  it("generates app port", async () => {
    const appTree = createTreeWithEmptyWorkspace({ layout: "apps-libs" });
    await appGenerator(appTree, options);

    const configFile = readJson(appTree, `${DEMO_APP_ROOT}/configs.json`);
    expect(configFile["appPort"]).toEqual(3001);
  });

  it("increments app port if another app already exists", async () => {
    const appTree = createTreeWithEmptyWorkspace({ layout: "apps-libs" });
    await appGenerator(appTree, options);

    const anotherAppOptions = { ...options, name: "another-app" };
    await appGenerator(appTree, anotherAppOptions);

    const configFile = readJson(appTree, "apps/another-app/configs.json");
    expect(configFile["appPort"]).toEqual(3002);
  });

  it("adds Vite configuration", async () => {
    const appTree = createTreeWithEmptyWorkspace({ layout: "apps-libs" });
    await appGenerator(appTree, options);

    const viteFile = appTree.read(`${DEMO_APP_ROOT}/vite.config.ts`)?.toString();
    expect(viteFile).toMatchSnapshot();
  });

  it("adds Vitest configuration", async () => {
    const appTree = createTreeWithEmptyWorkspace({ layout: "apps-libs" });
    await appGenerator(appTree, options);

    const vitestFile = appTree.read(`${DEMO_APP_ROOT}/vitest.config.ts`)?.toString();
    expect(vitestFile).toMatchSnapshot();
  });

  it("does not add E2E app", async () => {
    const appTree = createTreeWithEmptyWorkspace({ layout: "apps-libs" });
    await appGenerator(appTree, options);

    expect(appTree.exists(DEMO_APP_ROOT)).toBeTruthy();
    expect(appTree.exists(DEMO_APP_E2E_ROOT)).toBeFalsy();
  });

  describe("when custom tags are provided", () => {
    it("adds targets to the project configuration", async () => {
      const customDirOptions: AppGeneratorSchema = {
        ...options,
        tags: "tag1, tag2",
      };
      const appTree = createTreeWithEmptyWorkspace({ layout: "apps-libs" });
      await appGenerator(appTree, customDirOptions);

      const config = readProjectConfiguration(appTree, "demo");

      expect(config).toHaveProperty("tags", ["tag1", "tag2"]);
    });
  });

  describe("when custom directory is provided", () => {
    it("generates files in the custom directory", async () => {
      const customDirOptions: AppGeneratorSchema = {
        ...options,
        directory: "custom-demo",
        e2e: true,
      };
      const appTree = createTreeWithEmptyWorkspace({ layout: "apps-libs" });
      await appGenerator(appTree, customDirOptions);

      expect(appTree.exists(DEMO_APP_ROOT)).toBeFalsy();
      expect(appTree.exists("apps/custom-demo/demo")).toBeTruthy();
      expect(appTree.exists("apps/custom-demo/demo-e2e")).toBeTruthy();
    });
  });

  describe("when E2E app should be created", () => {
    const e2eOptions: AppGeneratorSchema = { ...options, e2e: true };

    it("generates E2E project configuration", async () => {
      const appTree = createTreeWithEmptyWorkspace({ layout: "apps-libs" });
      await appGenerator(appTree, e2eOptions);

      const config = readProjectConfiguration(appTree, "demo-e2e");

      config.targets && expect(Object.keys(config.targets)).toHaveLength(2);
      expect(config.targets).toHaveProperty("e2e");
      expect(config.targets).toHaveProperty("codegen");
    });

    it("adds Playwright configuration", async () => {
      const appTree = createTreeWithEmptyWorkspace({ layout: "apps-libs" });
      await appGenerator(appTree, e2eOptions);

      const playwrightFile = appTree.read(`${DEMO_APP_E2E_ROOT}/playwright.config.ts`)?.toString();
      expect(playwrightFile).toMatchSnapshot();
    });
  });

  describe("when Rest server should be added", () => {
    const serverOptions: AppGeneratorSchema = { ...options, restServer: true };

    it("generates database initializer script", async () => {
      const appTree = createTreeWithEmptyWorkspace({ layout: "apps-libs" });
      await appGenerator(appTree, serverOptions);

      expect(appTree.exists(`${DEMO_APP_ROOT}/server/initJsonDb.mts`)).toBeTruthy();
    });
  });
});
