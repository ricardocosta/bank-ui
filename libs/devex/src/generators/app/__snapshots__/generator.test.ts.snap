// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Generators / App > adds Vite configuration 1`] = `
"/// <reference types=\\"vite/client\\" />
import configs from './configs.json';
import react from '@vitejs/plugin-react-swc';
import { defineConfig } from 'vite';
import svgr from 'vite-plugin-svgr';
import tsconfigPaths from 'vite-tsconfig-paths';

// https://vitejs.dev/config/
export default defineConfig({
  plugins: [react(), svgr(), tsconfigPaths()],
  server: {
    port: configs.appPort,
  },
});
"
`;

exports[`Generators / App > adds Vitest configuration 1`] = `
"import viteConfig from './vite.config';
import { mergeConfig } from 'vite';
import { defineConfig } from 'vitest/config';

export default mergeConfig(
  viteConfig,
  defineConfig({
    test: {
      globals: true,
      environment: 'jsdom',
      setupFiles: './setupTests.ts',
      coverage: {
        reporter: ['lcov', 'text'],
      },
    },
  })
);
"
`;

exports[`Generators / App > when E2E app should be created > adds Playwright configuration 1`] = `
"import { APP_PORT } from './constants';
import type { PlaywrightTestConfig } from '@playwright/test';

const browser = process.env.BROWSER ?? 'chromium';
const os = process.env.OS ?? process.platform;

const config: PlaywrightTestConfig = {
  reporter: [
    [process.env.CI ? 'github' : 'list'],
    [
      'html',
      {
        open: 'on-failure',
        outputFolder: \`../../e2e/reports/apps/demo/\${os}/\${browser}\`,
      },
    ],
  ],

  testDir: 'tests',
  outputDir: \`../../e2e/results/apps/demo/\${os}/\${browser}\`,

  retries: 1,

  use: {
    baseURL: \`http://localhost:\${APP_PORT}\`,
    viewport: { width: 1280, height: 720 },
    ignoreHTTPSErrors: true,
    bypassCSP: true,

    trace: 'on-first-retry',
    screenshot: {
      mode: 'on',
      fullPage: true,
    },
  },

  webServer: {
    command: 'yarn nx run demo:dev',
    cwd: '../..',
    port: APP_PORT,
  },
};

export default config;
"
`;
